src/sources/mod.rs

```diff
@@ -356,15 +372,57 @@ pub(crate) trait EventDispatcher<Data> {
         data: &mut Data,
     ) -> crate::Result<PostAction>;
 
-    fn register(&self, poll: &mut Poll, token_factory: &mut TokenFactory) -> crate::Result<()>;
+    fn register(
+        &self,
+        poll: &mut Poll,
+        additional_lifetime_register: AdditionalLifetimeEventsRegister<'_>,
```

Comment:
What's the reason to pass this thing around instead of storing sources with lifetimes and sources without in separate buckets and simply pick based on the assoc constant? And have a way to iterate all the sources which simply takes both backets and `chain`s them?

Reply:
> And have a way to iterate all the sources which simply takes both backets and chains them?

I thought the operation which we perform isn't iterating, it's `.get`. If that's right, then the method would need to do two get based lookups for each event, which is potentially annoying. Additionally, due to the use of the slab datastructure, this would double the memory usage of the event loop, which isn't ideal.

Reply:
I mean, the buckets won't intersect at all. It's not like `Slab` will pre-allocate lots of storage just because you start using 2 of them?

You also maintain `HashSet` so it's it's hard to judge memory usage until you profile both approaches.

Reply:
Note, it is possible to iterate over the Slab contents, but this is discouraged because not efficient: https://docs.rs/slab/latest/slab/struct.Slab.html#method.iter

Reply:
I guess we could lose another bit in `Token` to storing whether the source needs additional lifecycle events. I'm not convinced that is better, though.

Reply:
So, as discussed on Matrix, the main point in favor of the current structure is that it allows to track when sources are disabled and reenabled, to avoid invoking `before_sleep()` on a disabled source. That seems like an important point to me.